### ISSUE TITLE
Circular Dependencies and Live Bindings

### ISSUE THREAD

**Issue by PepsRyuu** [PepsRyuu/nollup#19]
Nollup at the moment does not support circular dependencies. 

Webpack supports circular dependencies by implementing the "live binding" specification of ES Modules. It accomplishes this by replacing all import variables with an object, and accessing a special property on the object with a getter that gets the latest value.

```
// In the exported file, if we exported a variable called "number",
// it will add this to the exports instead of "number" directly.
Object.defineProperty(__webpack_exports__, "a", {
    configurable: false,
    enumerable: true,
    get: () => number
});

// Import variables are replaced to this renamed variable.
var __WEBPACK_IMPORTED_MODULE_0__a__ = __webpack_require__(0);

// All references to the imported variable must be replaced and does a get
// on the specially defined property. This property will return the latest value.
console.log(__WEBPACK_IMPORTED_MODULE_0__a__["a" /* number */]); 
```

I don't really want to implement this because of the following reasons:

* Significant complexity, would have to traverse the entire syntax tree, figured out if the imported variable was used, and replace it with the objects. This can introduce a significant performance penalty, and might not be entirely reliable unless all possibilities have been considered (eg. what if a local variable by the same name as an import is declared?)

* Breaks debugging symbols. By swapping out import variables with objects, it breaks debugging because the variables you see in the code are not the variables that are outputted. So if you want to watch variables, or test stuff in the console, you have to use the transpiled variable name instead which can be confusing.

* Live bindings is a rarely used feature. Should the live-binding specification be supported for a rare circular use case, and in those situations where it is a use case, is usually a code smell and can be refactored to not be circular? Should performance and debugging be hit where other workarounds may exist?

Would love to hear people's thoughts on this. I have no plans to implement this feature, but it is open to anyone who wants to try solve it.

**Comment by TheBosZ** [comment]
I'm absolutely okay with having to rewrite my code to resolve circular dependencies. In fact, in testing nollup, I had to do just that.

But I run into problems with libraries that I'm including. I can't really rewrite moment, for example. Yes, I could (and probably should) switch out moment for a better library or change how we're importing it, but that's not really a choice right now.

But that leads to something: I will **never** care about hot reloading Moment. It's a library that will only ever be either entirely included or entirely removed.

**Comment by ashubham** [comment]
I agree with @TheBosZ . We use `moment` too, and some other libs which have circular references. Is there a workaround for those ? Maybe a config item where you can specify modules to exclude from hot reload etc.

**Comment by PepsRyuu** [comment]
@ashubham Ideally moment needs to produce a proper bundled ES module. Currently what they have is a ```jsnext:main``` field in their ```package.json``` that points to their ```src``` directory. Instead of that, what they need to do is add the ```module``` field, and use Rollup to create a ```esm``` format file. Not entirely sure why they haven't done it yet, but it's only additional 2 lines of configuration from what I can tell. 

As for workarounds, there's currently two solutions:

* Use the ```external``` option for Moment, and use a ```<script>``` tag to import Moment. You might want to manually copy Moment into your ```public``` directory instead of relying on ```node_modules```. With this approach, you can still access Moment using ES import statements, but it will be available globally on window.

* Use ```rollup-plugin-commonjs```. The UMD bundle is available from Moment. I'm guessing you have ```rollup-plugin-node-resolve``` with ```jsnext``` set to true? If so, set it to false so that it will resolve to the UMD bundle.  With the CJS plugin, you can convert that UMD bundle into an ESM module. If you need to keep ```jsnext```, consider using ```rollup-plugin-alias``` and pointing ```moment``` to ```node_modules/moment/moment.js```.

**Comment by PepsRyuu** [comment]
@TheBosZ @ashubham 

I just had an idea on how I could implement circular dependencies, while still managing to preserve variable names. Uses a completely new approach. It's available in the following branch: https://github.com/PepsRyuu/nollup/tree/LiveBindingExperiment . There's a folder called ```example-circular``` that successfully imports the latest version of ```moment```.

This is still very much experimental. There's no guarantee on when/if this will go to master. Still have to do a lot of testing with it. I'm sharing at the moment so I can get some feedback, to make sure I'm not missing anything fundamental. I'd greatly appreciate it if you could give it a try in your projects if you get the chance.

**Behind the Scenes Explanation**

```
1: function (__r__, __d__, __e__) {
    var _i0 = {};
    var imported_variable_name;

    __d__(function () {
        imported_variable_name = _i0.imported_variable_name;
    }, function (require, module) {
        "use strict";
        eval("_e__('default', imported_variable_name ('some_argument'));");
    });

    _i0 = __r__(0);
}
```

The idea behind this approach, is to create an outer scope, with the imported variable names. The first function in the ```__d__``` call is the binder. Whenever ```__e__``` is called, it calls the binder of the parent modules that imported that module, simulating live bindings. The ```__r__``` function is similar to require, in that it will check child imports, but it doesn't evaluate them immediately. With this approach, I can "pre-parse" modules and lazy-evaluate the exports afterwards. This is important because it lets us know who all of the parents are prior to exporting.

To solve circular dependencies, when exports are being evaluated, I immediately mark a module as "resolving". Before calling the implementation of a module, I check it's dependencies recursively. If I hit a dependency that's already marked as "resolving" but not "resolved", it must be circular. At that point the remaining non-"resolved" non-"resolving" dependencies are checked until none are left and the module is immediately evaluated and set to "resolved". 

Theoretically this same approach can be used to solve the live binding problem. Each time an exported variable has been re-assigned, the ```__e__``` function can be called, and it will update all of the parents binders. I haven't implemented that part yet as it requires a lot of work and could still come with a performance penalty as you would still need to parse the entire tree and see if a variable was used and the different ways in which variables can be re-assigned.

**Comment by PepsRyuu** [comment]
Released Circular Dependency support in 0.6.0.

Live-Bindings hasn't been implemented, but the same approach can be used for it if needed. Very reluctant to implement that as it would require significantly complex parsing. For the time being though, closing this issue as circulars as originally requested is working now! :)
